src/ai/backend/agent/server.py

```diff
@@ -346,6 +351,39 @@ async def gather_hwinfo(self) -> Mapping[str, HardwareMetadata]:
     async def ping_kernel(self, kernel_id: str):
         log.debug("rpc::ping_kernel({0})", kernel_id)
 
+    @rpc_function
+    @collect_error
+    async def sync_kernel_registry(
+        self,
+        raw_kernel_session_ids: Iterable[tuple[str, str]],
+    ) -> None:
+        kernel_session_ids = [
+            (KernelId(UUID(raw_kid)), SessionId(UUID(raw_sid)))
+            for raw_kid, raw_sid in raw_kernel_session_ids
+        ]
+        for kid, sid in kernel_session_ids:
+            if kid not in self.agent.kernel_registry:
+                # produce KernelTerminatedEvent
+                await self.agent.produce_event(
+                    KernelTerminatedEvent(
+                        kid,
+                        sid,
+                        reason=KernelLifecycleEventReason.ALREADY_TERMINATED,
+                    )
+                )
+
+        kernel_ids = {kern_id for kern_id, sess_id in kernel_session_ids}
+        for kid, kernel in self.agent.kernel_registry.items():
+            if kid not in kernel_ids:
+                # destroy kernel
+                await self.agent.inject_container_lifecycle_event(
+                    kid,
+                    kernel.session_id,
+                    LifecycleEvent.DESTROY,
+                    KernelLifecycleEventReason.NOT_FOUND_IN_MANAGER,
+                    suppress_events=True,
+                )
```

Comment:
How about saving `self.agent.kernel_registry` into the `last_registry` file after sync?

Reply:
good point. then the agent will not load the "broken" registry pickle